txstratum/protocol.py

```diff
@@ -0,0 +1,358 @@
+        self.current_weight: float = self.INITIAL_WEIGHT
+        self.hashrate_ghs: float = 0.0
+
+        self._submitted_work: List[SubmittedWork] = []
+
+        self._authorized: bool = False
+        self._subscribed: bool = False
+        self.started_at: float = 0.0
+        self.estimator_task: Optional[Periodic] = None
+        self.refresh_job_task: Optional[Periodic] = None
+
+        # For testing only.
+        self._update_job_timestamp: bool = True
+
+        self.last_submit_at: float = 0.0
+        self.completed_jobs: int = 0
+        self.txs_solved: int = 0
+        self.blocks_found: int = 0
+        self.supported_methods: Dict[str, Callable[[Any, JSONRPCId], None]] = {
+            'subscribe': self.method_subscribe,
+            'mining.subscribe': self.method_subscribe,
+            'login': self.method_subscribe,
+
+            'authorize': self.method_authorize,
+            'mining.authorize': self.method_authorize,
+
+            'submit': self.method_submit,
+            'mining.submit': self.method_submit,
+
+            'configure': self.method_configure,
+            'mining.configure': self.method_configure,
+
+            'multi_version': self.method_multi_version,
+            'mining.multi_version': self.method_multi_version,
+
+            'mining.extranonce.subscribe': self.method_extranonce_subscribe,
+        }
+
+    def connection_made(self, transport: asyncio.BaseTransport) -> None:
+        """Set things up after a new connection is made.
+
+        See: https://docs.python.org/3/library/asyncio-protocol.html#asyncio.BaseProtocol.connection_made
+        """
+        self.transport = cast(asyncio.Transport, transport)
+        self.manager.add_connection(self)
+        self.log = self.log.bind(miner_id=self.miner_id)
+        self.log.debug('connection made')
+
+    def connection_lost(self, exc: Optional[Exception]) -> None:
+        """Clean up after a connection is lost.
+
+        See: https://docs.python.org/3/library/asyncio-protocol.html#asyncio.BaseProtocol.connection_lost
+        """
+        self.log.debug('connection lost', exc=exc)
+        if self._subscribed:
+            self.log.info('Miner exited')
+        assert self.miner_id is not None
+        self.manager.remove_connection(self)
+        if self.estimator_task:
+            asyncio.ensure_future(self.estimator_task.stop())
+        if self.refresh_job_task:
+            asyncio.ensure_future(self.refresh_job_task.stop())
+
+    def start_estimator(self) -> None:
+        """Start periodic estimator task."""
+        if self.estimator_task is None:
+            self.last_reduced = time.time()
+            # Start estimator periodic task.
+            self.estimator_task = Periodic(self.estimator_loop, self.ESTIMATOR_LOOP_INTERVAL)
+            asyncio.ensure_future(self.estimator_task.start())
+            # Start fresh job periodic task.
+            self.refresh_job_task = Periodic(self.refresh_job, self.JOB_UPDATE_INTERVAL)
```

Comment:
This will run every 2 seconds, right? There are 2 implications:
1. For blocks, we don't need to update the job every 2 seconds. I'm not sure if it impacts performance (and in the end hash rate estimation), but as we're not concerned with mining blocks here, it's probably not a problem;
2. Ideally, we should start the 2s timer when we send the new tx job to the miner. Since this runs every 2 seconds regardless of when the job starts, it might happen that it'll update the job after it's been running for a few milliseconds. After this first one, the job send interval should "sync" with this periodic task, so maybe there's not much impact.

Reply:
1) It shouldn't impact performance because we don't send `clean=True`. So, the client can decide the best cycle to update the job.

2) As we are using `clean=False`, it seems ok. It might happen that a miner will receive a job and an update just after, but it shouldn't affect the overall performance.